internal/controller/cleanup.go

```diff
@@ -0,0 +1,160 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	monitoringv1 "github.com/coreos/prometheus-operator/pkg/apis/monitoring/v1"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// getPromRulefromAbsentRuleGroup takes the name of a RuleGroup name that holds
+// absent alerts and returns the name of the origin PrometheusRule that holds
+// the corresponding alert definitions, and a bool which is false if the name
+// can't be determined.
+//
+// Absent alert RuleGroups names have the format:
+//   originPrometheusRuleName/RuleGroupName
+func getPromRulefromAbsentRuleGroup(group string) (string, bool) {
+	sL := strings.Split(group, "/")
+	if len(sL) != 2 {
+		return "", false
+	}
+	return sL[0], true
+}
+
+//nolint:gocognit
+// cleanUpOrphanedAbsentAlertsNamespace deletes orphaned absent alerts across a
+// cluster.
+func (c *Controller) cleanUpOrphanedAbsentAlertsCluster() error {
+	// Get names of all PrometheusRules that exist in the informer's cache:
+	//   map of namespace to map[promRuleName]bool
+	promRules := make(map[string]map[string]bool)
+	objs := c.promRuleInformer.GetStore().List()
+	for _, v := range objs {
+		pr, ok := v.(*monitoringv1.PrometheusRule)
+		if !ok {
+			continue
+		}
+		ns := pr.GetNamespace()
+		if _, ok = promRules[ns]; !ok {
+			promRules[ns] = make(map[string]bool)
+		}
+		promRules[ns][pr.GetName()] = true
+	}
+
+	cleanup := make(map[string]map[string]struct{})
+	for namespace, promRuleNames := range promRules {
+		// Get all absentPrometheusRules for this namespace.
+		prList, err := c.promClientset.MonitoringV1().PrometheusRules(namespace).
+			List(context.Background(), metav1.ListOptions{LabelSelector: labelOperatorManagedBy})
+		if err != nil {
+			return errors.Wrap(err, "could not list AbsentPrometheusRules")
+		}
+
+		// Check if there are any alerts in those absentPrometheusRules that
+		// don't belong to any PrometheusRule in promRuleNames.
+		for _, pr := range prList.Items {
+			for _, g := range pr.Spec.Groups {
+				n, ok := getPromRulefromAbsentRuleGroup(g.Name)
+				if !ok {
+					continue
+				}
+				if promRuleNames[n] {
+					continue
+				}
+				if _, ok = cleanup[namespace]; !ok {
+					cleanup[namespace] = make(map[string]struct{})
+				}
+				cleanup[namespace][n] = struct{}{}
+			}
+		}
+	}
+
+	for ns, m := range cleanup {
+		for pr := range m {
+			if err := c.cleanUpOrphanedAbsentAlertsNamespace(pr, ns); err != nil {
+				return err
+			}
+		}
+	}
```

Comment:
I don't see any reason for collecting rules into the `cleanup` map first, then doing the cleanup in a second set of loops. It should be safe to just call `cleanUpOrphaned...` directly in the first set of loops, no?